### PR TITLE
start dedicated info modules

### PR DESCRIPTION
**Pull request by evgeni**
starting off with `host` as that's rather simple.

next on my list would be `repository` which is a bit more involved, but I hope to be able to abstract most of it into the helper

### CODE REVIEW

**Comment by evgeni** [comment]
@ehelms as discussed the other day :)

File: plugins/modules/host_info.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+def main():
+    module = ForemanHostInfo()
+
+    with module.api_connection():
+        module.run()
+
+
+if __name__ == '__main__':
+    main()
```
Comment (evgeni) [inline plugins/modules/host_info.py:81]: so much real code… ;)

Comment (mdellweg) [reply plugins/modules/host_info.py:81]: mostly blank lines anyway. ;)

File: plugins/modules/host_info.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  type: list
+  elements: dict
+'''
+
+from ansible_collections.theforeman.foreman.plugins.module_utils.foreman_helper import (
+    ForemanInfoAnsibleModule,
+)
+
+
+class ForemanHostInfo(ForemanInfoAnsibleModule):
```
Comment (mdellweg) [inline plugins/modules/host_info.py:69]: If you called this a `ForemanHostInfoModule` the resource lookup was broken, right?

Comment (evgeni) [reply plugins/modules/host_info.py:69]: Yeah, then I'd need to adjust `entity_name_from_class` ;)

We could, tho, just /me being lazy.

File: plugins/modules/host_info.py (hunk 1 above)
Comment (evgeni) [reply plugins/modules/host_info.py:81]: I wondered about folding `with connection: run` into an `autorun` method?

Comment (evgeni) [reply plugins/modules/host_info.py:81]: so then the whole main would become:
```
def main():
    ForemanHostInfo().autorun()
```

File: plugins/module_utils/foreman_helper.py [hunk 3]
```diff
@@ -1399,11 +1431,9 @@ class KatelloAnsibleModule(KatelloMixin, ForemanAnsibleModule):
     pass
 
 
-class KatelloEntityAnsibleModule(KatelloMixin, ForemanEntityAnsibleModule):
+class KatelloScopedMixin(KatelloMixin):
```
Comment (evgeni) [inline plugins/module_utils/foreman_helper.py:1452]: I wonder if we really need this, or if the scope enforcement should happen in `KatelloMixin` directly.

File: plugins/modules/host_info.py (hunk 1 above)
Comment (mdellweg) [reply plugins/modules/host_info.py:81]: I don't think it's too much boilerplate, and it keeps the simple modules in the same stucture as the ones where we need some manipulation between `api_connection` and `run`.
I think it's a good thing, that you can see this shared inheritance.

File: plugins/modules/host_info.py (hunk 2 above)
Comment (mdellweg) [reply plugins/modules/host_info.py:69]: `ForemanHostInfomodule` could work, but maybe that introduces even more confusion.

File: plugins/module_utils/foreman_helper.py (hunk 3 above)
Comment (mdellweg) [reply plugins/module_utils/foreman_helper.py:1452]: Maybe yes, i guess, you need to look into an non entity katello module to decide.

File: tests/test_playbooks/host_info.yml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+# TBD
```
Comment (ehelms) [inline tests/test_playbooks/host_info.yml]: Is this TBD in this PR or a follow up?

Comment (evgeni) [reply tests/test_playbooks/host_info.yml]: This PR, I waa just lazy ;)

**Comment by evgeni** [comment]
So I've added tests, and they led me to some design-y questions…

Currently, the modules return their data in the `resources` attribute (so if you register the host info as `host`, to access it you need to do `host.resources…`), which is an inheritance from the `resource_info` module. But the data structure returned will differ, based on `name` (one item) vs `search` (array of items). Which leads me to the following questions:
* Should the attribute be called `<whatever_resource>` instead? We know the resource name after all
* Should the attribute vary and be `<whatever_resource>` for a single name result, vs `<whatever_resource>s` for the search result? We know both singular and plural names too
* Should the returned data structure be always an array?

File: tests/test_playbooks/host_info.yml [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        validate_certs: "{{ foreman_validate_certs }}"
+        name: "test-host.{{ host.domain }}"
+        organization: "{{ host.organization }}"
+        location: "{{ host.location }}"
+      register: host_info
+    - name: check host details
+      assert:
+        that:
+          - host_info['resources']['name'] == "test-host.{{ host.domain }}"
+          - host_info['resources']['domain_name'] == host.domain
```
Comment (evgeni) [inline tests/test_playbooks/host_info.yml]: this is the part I mostly wonder about

File: tests/test_playbooks/host_info.yml [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        validate_certs: "{{ foreman_validate_certs }}"
+        search: "name = test-host.{{ host.domain }}"
+        organization: "{{ host.organization }}"
+        location: "{{ host.location }}"
+      register: host_info
+    - name: check host details
+      assert:
+        that:
+          - host_info['resources'][0]['name'] == "test-host.{{ host.domain }}"
+          - host_info['resources'][0]['domain_name'] == host.domain
```
Comment (evgeni) [inline tests/test_playbooks/host_info.yml]: vs this

**Comment by mdellweg** [comment]
I think i prefer to be specific and use `<whatever_resource>(s)` accordingly. So we can expect
```
result => {
"state": "OK",
"host": {
  "name": "whoever.whereev.er",
  ...
}
}
```
Am i getting this right?

**Comment by evgeni** [comment]
for `name: host.example.com` you'd get:

```
ok: [tests] => {
    "changed": false,
    "host": {
        "id": 12,
        "name": "host.example.com",
        …
     }
}
```

and for for `search: name = host.example.com` you'd get:

```
ok: [tests] => {
    "changed": false,
    "hosts": [
        {
          "id": 12,
          "name": "host.example.com",
          …
        }
     ]
}
```

**Comment by evgeni** [comment]
Updated to reflect the latest discussion results.

**Comment by evgeni** [comment]
~~Should add docs in `RETURNS`…~~ done

File: plugins/module_utils/foreman_helper.py [hunk 7]
```diff
@@ -1287,6 +1287,40 @@ def entity_name_from_class(self):
...
+
+    def run(self, **kwargs):
+        """
+        lookup entities
+        """
+        self.auto_lookup_entities()
+
+        resource = self.foreman_spec['entity']['resource_type']
+
+        if 'name' in self.foreman_params:
```
Comment (ehelms) [inline plugins/module_utils/foreman_helper.py:1313]: Checking on my pythonic learning, should this be a try-rescue block for a KeyError ?

**Review by ehelms — APPROVED** [review]
Overall looks good to me, I just had the one question to help with my python learning.

File: plugins/module_utils/foreman_helper.py (hunk 7 above)
Comment (mdellweg) [reply plugins/module_utils/foreman_helper.py:1313]: This looks like politely asking if a KeyError would be expected.

Comment (evgeni) [reply plugins/module_utils/foreman_helper.py:1313]: a try/except block? no-ish.
the code doesn't access `foreman_params['name']`, so there is nothing that would raise the `KeyError`.
(well, `lookup_entity` might at some point, but that's weird)

File: plugins/doc_fragments/foreman.py [hunk 8]
```diff
@@ -325,3 +325,35 @@ class ModuleDocFragment(object):
...
+
+    INFOMODULE = '''
+options:
+  name:
+    description:
+      - Name of the resource to fetch information for.
+      - Mutually exclusive with I(search).
+    required: false
+    type: str
+  location:
```
Comment (mdellweg) [inline plugins/doc_fragments/foreman.py:337]: Do all resourced have taxonomies?
Well probably all that we have infomodules for at this point. Which is good enough for now.

File: plugins/doc_fragments/foreman.py [hunk 9]
```diff
@@ -325,3 +325,35 @@ class ModuleDocFragment(object):
...
+      - Search query to use
+      - If None, and I(name) is not set, all resources are returned.
+      - Mutually exclusive with I(name).
+    type: str
+'''
+
+    KATELLOINFOMODULE = '''
+options:
+  organization:
+    required: true
```
Comment (mdellweg) [inline plugins/doc_fragments/foreman.py:358]: :+1:

**Review by mdellweg — APPROVED** [review]

File: plugins/doc_fragments/foreman.py (hunk 8 above)
Comment (evgeni) [reply plugins/doc_fragments/foreman.py:337]: Yeah, `setting` is not taxonomied, but I've left that to future-me to fix :)

**Pull request merged**